### DISCUSSION TITLE
Required fields in conditionally hidden forms

### DISCUSSION THREAD

**Discussion by boosh** [jrief/django-formset#117] [Q&A]
I have some conditionally hidden fieldsets, but fields within them are required. When the form is submitted, validation fails. 

The same thing happens in [the docs](https://django-formset.fly.dev/conditionals/#conditional-fieldset). With the "I'm not a customer" checkbox selected, the response comes back with a validation error complaining about a hidden missing required field:

```
{
  "request": {
    "formset_data": {
      "customer": {
        "recipient": "",
        "email": ""
      },
      "register": {
        "no_customer": "on"
      }
    }
  },
  "response": {
    "customer": {
      "recipient": [
        "This field is required."
      ]
    },
    "register": {}
  }
}
```

I was hoping the library would automatically validate empty fields in hidden forms but that doesn't seem to be the case. What's the best way of handling this? Are we to just implement a custom `full_clean()` method on forms allowing empty hidden required fields, or is there another way or an integrated feature I'm missing?

Thanks

**Comment by jrief** [comment]
This kind of "business logic" must be implemented by the `full_clean()`-method of your form collection. This also is documented, please check the [note](https://django-formset.fly.dev/conditionals/) in the cited docs.

I actually don't have any good idea on how to implement this. `df-show` and `df-hide` just change the `display` property of the referenced element. Even if I would remove the submission data of the hidden elements, the Django form validation would still complain.

Maybe you have a better idea.

**Reply by boosh** [reply]
Ah I hadn't seen that note. I suppose there are definitely cases where a field should be required if a condition is met. E.g. in the example I linked to from the docs, it's right that recipient is a required field, but only if "I'm not a customer" is blank.

I wondered whether the form classes could traverse their children, evaluate show/hide conditions and if e.g. a show condition is false, just ignore empty required fields (and vice versa). That's basically all I'll be doing manually. 

I'm not sure how difficult that'd be to implement though - you know the code better than I do. Conceptually it's just some iteration and probably passing in the whole data dict to evaluate conditions in each form, then dynamically setting `self.fields[<name>].required = False` I guess...

**Comment by jrief** [comment]
Recently I added the optional attribute `df-require` to input fields. Have you been able to test it?

**Reply by boosh** [reply]
No I'm just handling it with a custom clean method for now. Not sure when I'll get chance to test this tbh.
